File: Binary Search/3350. Adjacent Increasing Subarrays Detection II/solution.py
class Solution:
    def maxIncreasingSubarrays(self, nums: List[int]) -> int:
        n = len(nums)

        max_length = [0] * (n+1)
        for i in range(n):
            if nums[i] > nums[i-1]:
                max_length[i+1] = max_length[i] + 1
            else:
                max_length[i+1] = 1

        def check(k):
            for i in range(1, n+1):
                if i-k >= 0 and max_length[i] >= k and max_length[i-k] >= k:
                    return True
            return False
        
        l, r = 1, n//2
        while l < r:
            mid = r - (r-l)//2
            if check(mid):
                l = mid
            else:
                r = mid-1
        return l

File: Binary Search/3350. Adjacent Increasing Subarrays Detection II/solution.cpp
/**
2 cases:
    1. one increasing subarray: k = length / 2
    2. two adjacent subarray: k = min(length1, length2) ex. X X X X X X X {X X X} {Y Y Y}
*/
class Solution {
    
public:
    int maxIncreasingSubarrays(vector<int>& nums) {
        int n = nums.size();
        
        int ans = 1;
        int prev = 0;
        int length = 1;
        for(int i = 1; i < n; i++) {
            if(nums[i - 1] < nums[i]) {
                length++;
            } else {
                prev = length;
                length = 1;
            }
            ans = max(ans, max(length / 2, min(length, prev)));
        }
            
       
        return ans;

    }
};